File: CDAC/DS_Team_6/DoubleEndedQueue.cpp
#include<iostream>
using namespace std;
class Queue{
    private:
        int *arr;
        int size;
        int front;
        int rear;
    public:
        Queue(int size){
            arr = new int[size];
            this->size = size;
            front = -1;
            rear = -1;
        }

        bool isEmpty(){
            if(front == -1 || front > rear){
                return true;
            }else{
                return false;
            }
        }

        bool isFull(){
            if(front == 0 && rear == size - 1){
                return true;
            }else{
                return false;
            }
        }

        void enqueueAtEnd(int data){
            if(!isFull()){
                if(rear >= size-1){
                    cout<<"Can't enqueue at the end!!!"<<endl;
                }else{
                    if(front == -1)
                        front++;
                    rear++;
                    arr[rear] = data;
                    cout<<data<<" enqueued successfully!!!"<<endl;
                }
                
            }else{
                cout<<"Queue is full!!!"<<endl;
            }
        }

        void dequeueFromStart(){
            if(!isEmpty()){
                cout<<"Dequeued element is "<<arr[front++]<<endl;
            }else{
                cout<<"Queue is empty!!!"<<endl;
            }
        }

        void enqueueAtStart(int data){
            if(!isFull()){
                if(front == -1){
                    front++;
                    rear++;
                    arr[front] = data;
                    cout<<data<<" enqueued successfully!!!"<<endl;
                }else if(front > 0){
                    front--;
                    arr[front] = data;
                    cout<<data<<" enqueued successfully!!!"<<endl;
                }else{
                    cout<<"Can't enqueue from start!!!"<<endl;
                }
            }else{
                cout<<"Queue is full!!!"<<endl;
            }
        }

        void dequeueFromEnd(){
            if(!isEmpty()){
                cout<<"Dequeued element is "<<arr[rear--]<<endl;
            }else{
                cout<<"Queue is empty!!!"<<endl;
            }
        }

        void display(){
            if(!isEmpty()){
                cout<<"Elements in queue are ";
                for(int i=front;i<=rear;i++){
                    cout<<arr[i]<<" ";
                }
                cout<<endl;
            }else{
                cout<<"Queue is empty!!!"<<endl;
            }
        }
};
int main(){
    int size,data,choice;
    cout<<"Enter the size of data: ";
    cin>>size;
    Queue q(size);
    while(1){
        cout<<"1. Enqueue at start\n2. Dequeue from end\n3. Enqueue at end\n4. Dequeue from start\n5. Display\n6. Exit"<<endl;
        cout<<"Enter your choice: ";
        cin>>choice;
        switch(choice){
            case 1:
                cout<<"Enter the data: ";
                cin>>data;
                q.enqueueAtStart(data);
                break;
            case 2:
                q.dequeueFromEnd();
                break;
            case 3:
                cout<<"Enter the data: ";
                cin>>data;
                q.enqueueAtEnd(data);
                break;
            case 4:
                q.dequeueFromStart();
                break;
            case 5: 
                q.display();
                break;
            case 6:
                exit(0);
            default:
                cout<<"Enter the correct choice!!!"<<endl;
        }
    }
}

File: CDAC/edac_data_structure/src/edac_data_structure/DoubleEndedQueue.java
package edac_data_structure;

import java.util.Scanner;

public class DoubleEndedQueue {

	private int[] arr;
	private int size;
	private int front;
	private int rear;
	
	DoubleEndedQueue(){
		arr = new int[0];
		size=0;
		front = rear = -1;
	}
	
	DoubleEndedQueue(int size){
		arr = new int[size];
		this.size = size;
		front = rear = -1;
	}
	
	void enqueueAtFront(int data) {
		if(front == 0 && rear == size-1) {
			System.out.println("Queue is full!!!");
		}
		else if(front == -1 && rear == -1) {
			front++;
			rear++;
			arr[rear]=data;
			System.out.println(data+" added successfully!!!");
		}else {
			if(front>0) {
				front--;
				arr[front]=data;
				System.out.println(data+" added successfully!!!");
			}else {
				System.out.println(data+ " can not be added!!!");
			}
		}
	}
	
	void enqueueAtRear(int data) {
		if(front == 0 && rear == size-1) {
			System.out.println("Queue is full!!!");
		}
		else if(front == -1 && rear == -1) {
			front++;
			rear++;
			arr[rear]=data;
			System.out.println(data+" added successfully!!!");
		}else {
			if(rear<size-1) {
				rear++;
				arr[rear]=data;
				System.out.println(data+" added successfully!!!");
			}else {
				System.out.println(data+ " can not be added!!!");
			}
		}
	}
	
	void dequeueFromFront() {
		if(front == -1 && rear == -1)
			System.out.println("Queue is empty!!!");
		else if(front>rear) {
			System.out.println("Queue is empty!!!");
		}else {
			System.out.println(arr[front]+" removed successfully!!!");
			front++;
		}
	}
	
	void dequeueFromRear() {
		if(front == -1 && rear == -1)
			System.out.println("Queue is empty!!!");
		else if(front>rear) {
			System.out.println("Queue is empty!!!");
		}else {
			System.out.println(arr[rear]+" removed successfully!!!");
			rear--;
		}
	}
	
	void print() {
		if(front == -1 && rear == -1)
			System.out.println("Queue is empty!!!");
		else if(front>rear) {
			System.out.println("Queue is empty!!!");
		}else {
			for(int i=front;i<=rear;i++) {
				System.out.print(arr[i]+" ");
			}
			System.out.println();
		}
	}
	
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		System.out.println("Enter the size of queue:");
		int size = sc.nextInt();
		DoubleEndedQueue q = new DoubleEndedQueue(size);
		int data,choice=0;
		
		while(choice != 6) {
			System.out.println("1. Enqueue at front\n2. Enqueue at rear\n3. Dequeue from front\n4. Dequeue from rare\n5. Print\n6. Exit");
			System.out.println("Enter your choice:");
			choice = sc.nextInt();
			switch(choice) {
			case 1:
				System.out.println("Enter the data:");
				data = sc.nextInt();
				q.enqueueAtFront(data);
				break;
			case 2:
				System.out.println("Enter the data:");
				data = sc.nextInt();
				q.enqueueAtRear(data);
				break;
			case 3:
				q.dequeueFromFront();
				break;
			case 4:
				q.dequeueFromRear();
				break;
			case 5:
				q.print();
				break;
			case 6:
				System.exit(0);
				break;
			default:
				System.out.println("Enter valid choice");
			}
		}
		sc.close();
	}
}
